client/asset/eth/eth.go

```diff
@@ -535,165 +540,151 @@ func decodeCoinID(coinID []byte) (*coin, error) {
 
-	currentlyLocking := make(map[string]bool)
-	var amountToLock uint64
-	for _, coin := range coins {
-		hexID := hex.EncodeToString(coin.ID())
-		if _, ok := eth.lockedFunds[hexID]; ok {
-			return fmt.Errorf("cannot lock funds that are already locked: %v", hexID)
-		}
-		if _, ok := currentlyLocking[hexID]; ok {
-			return fmt.Errorf("attempting to lock duplicate coins: %v", hexID)
-		}
-		currentlyLocking[hexID] = true
-		amountToLock += coin.Value()
-	}
-
-	balance, err := eth.balanceImpl()
+	balance, err := eth.balance()
 	if err != nil {
 		return err
 	}
-
-	if balance.Available < amountToLock {
+	if balance.Available < amt {
 		return fmt.Errorf("currently available %v < locking %v",
-			balance.Available, amountToLock)
+			balance.Available, amt)
 	}
 
-	for _, coin := range coins {
-		eth.lockedFunds[hex.EncodeToString(coin.ID())] = coin.Value()
-	}
+	eth.acct.locked += amt
 	return nil
 }
 
 // lockFunds removes coins from the map of locked funds.
-func (eth *ExchangeWallet) unlockFunds(coins asset.Coins) error {
-	eth.lockedFundsMtx.Lock()
-	defer eth.lockedFundsMtx.Unlock()
-
-	currentlyUnlocking := make(map[string]bool)
-	for _, coin := range coins {
-		hexID := hex.EncodeToString(coin.ID())
-		if _, ok := eth.lockedFunds[hexID]; !ok {
-			return fmt.Errorf("cannot unlock coin ID that is not locked: %v", coin.ID())
-		}
-		if _, ok := currentlyUnlocking[hexID]; ok {
-			return fmt.Errorf("attempting to unlock duplicate coins: %v", hexID)
-		}
-		currentlyUnlocking[hexID] = true
-	}
+func (eth *ExchangeWallet) unlockFunds(amt uint64) error {
+	eth.acct.lockedMtx.Lock()
+	defer eth.acct.lockedMtx.Unlock()
 
-	for _, coin := range coins {
-		delete(eth.lockedFunds, hex.EncodeToString(coin.ID()))
+	if eth.acct.locked < amt {
+		return fmt.Errorf("currently locked %v < unlocking %v",
+			eth.acct.locked, amt)
 	}
 
+	eth.acct.locked -= amt
 	return nil
 }
 
 // Swap sends the swaps in a single transaction. The Receipts returned can be
 // used to refund a failed transaction. The Input coins are manually unlocked
 // because they're not auto-unlocked by the wallet and therefore inaccurately
 // included as part of the locked balance despite being spent.
+//
+// NOTE: we must also emulate "change", which becomes the new amount locked by
+// parent order that the consumer will unlock with ReturnCoins when the order is
+// retired (no more swaps and no longer booked). The change amount is the input
+// amount minus the actual amount *spent* by this swap txn.
 func (*ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin, uint64, error) {
+	// TODO: remember to unlockFunds when ETH is sent to a swap contract.
 	return nil, nil, 0, asset.ErrNotImplemented
 }
 
 // Redeem sends the redemption transaction, which may contain more than one
 // redemption.
 func (*ExchangeWallet) Redeem(form *asset.RedeemForm) ([]dex.Bytes, asset.Coin, uint64, error) {
+	// TODO: remember to unlockFunds for the reserved redeem gas amount.
```

Comment:
When would those funds get locked in the first place?

Reply:
This is an open question, but a as a thought I made a comment https://github.com/decred/dcrdex/pull/1289/files#diff-671406d18db2f8e1a281545be05fcdefd789b63614398d678d38e2b2814cf788R549 `FundBuyOrder`